Misc Fixes vor Windows / CUDA 10.0

jvschroeder opened the pull request:
Fixed some isssues I encountered while trying to compile the examples on Windows using CUDA 10.0

eyalroz commented on CMakeLists.txt:
```diff
@@ -119,7 +119,7 @@ add_executable(version_management EXCLUDE_FROM_ALL examples/by_runtime_api_modul
 add_executable(device_management EXCLUDE_FROM_ALL examples/by_runtime_api_module/device_management.cpp)
 add_executable(execution_control EXCLUDE_FROM_ALL examples/by_runtime_api_module/execution_control.cu)
-target_compile_options(execution_control PRIVATE -rdc true)
+#target_compile_options(execution_control PRIVATE -rdc true)
```
IIRC, this is necessary on non-Windows machines, I can't just remove it.

jvschroeder replied:
Ok - I'll make the change conditional then.

eyalroz commented on cmake/Modules/HandleCUDAComputeCapability.cmake:
```diff
@@ -8,24 +8,44 @@
-		set(ENV{CMAKE_CXX_COMPILER} "${CMAKE_CXX_COMPILER}")
-		execute_process(COMMAND bash -c "${CMAKE_CURRENT_SOURCE_DIR}/scripts/get_cuda_sm.sh"  OUTPUT_VARIABLE CUDA_TARGET_COMPUTE_CAPABILITY_) 
+	if(WIN32)
+		find_package(CUDA)
```
CUDA is now an integral part of CMake, we don't use find_package on it. All CUDA-related variables are built-in.

eyalroz commented on cmake/Modules/HandleCUDAComputeCapability.cmake:
```diff
@@ -8,24 +8,44 @@
+	if(WIN32)
+		find_package(CUDA)
+		if (CUDA_FOUND)
+			set(CUDAFILE ${CMAKE_CURRENT_SOURCE_DIR}/examples/check_cuda.cu)
```
It seems you've developed a facility for detecting the CUDA SM on Windows. Please:

1. Make that a separate PR
2. Put it in a CMake module
3. If you can spare the time - See how much of the Unix-oriented shell script can be discarded in favor of what you do in CMake, so that the code paths will not look so different.

jvschroeder replied:
regarding 3. - I do not see a reason why we couldn't use the CMake version on Unix as well. - I simply didn't delete it and made the Windows stuff conditional since I can't test if it has exactly the same behaviour.

eyalroz commented on cmake/Modules/HandleCUDAComputeCapability.cmake:
```diff
@@ -8,24 +8,44 @@
-		set(CUDA_TARGET_COMPUTE_CAPABILITY_ $ENV{CUDA_SM})
-	endif()
+		if("$ENV{CUDA_SM}" STREQUAL "")
+			set(ENV{CUDA_INCLUDE_DIRS} "${CUDA_INCLUDE_DIRS}")
```
Again, there are internal CMake variables for that.

jvschroeder replied:
set(ENV stuff is from the current (unix) version of the module

eyalroz commented on examples/by_runtime_api_module/device_management.cpp:
```diff
@@ -62,7 +62,7 @@ int main(int argc, char **argv)
 	// ---------------------------
 
-	auto max_registers_per_block = device.get_attribute(cudaDevAttrMaxRegistersPerBlock);
+	auto max_registers_per_block = device.get_attribute(static_cast<cuda::device::attribute_t>(cudaDevAttrMaxRegistersPerBlock));
```
Ah, it looks like the return type of get_attribute needs to change, there shouldn't be a cast here.

jvschroeder replied:
I suspect as much, but since I currently don't have the opportunity to test this on a non-Windows machine I don't know if changing the return type would be ok.

eyalroz commented on examples/check_cuda.cu:
```diff
@@ -0,0 +1,26 @@
+#include <stdio.h>
```
Again, move this to a separate PR.

eyalroz commented on examples/modified_cuda_samples/vectorAdd/vectorAdd.cu:
```diff
@@ -59,7 +59,7 @@ int main(void)
 	cuda::launch(
 		vectorAdd,
-		{ blocksPerGrid, threadsPerBlock },
+		{ cuda::grid_block_dimensions_t(blocksPerGrid), cuda::grid_block_dimensions_t(threadsPerBlock) },
```
1. Why did you need this?
2. The two parameters have different types.

jvschroeder replied:
If I don't have this then I get the following errors

> ./cuda-api-wrappers/examples/modified_cuda_samples/vectorAdd/vectorAdd.cu(60): error C23
> 97: conversion from 'int' to 'cuda::grid_dimensions_t' requires a narrowing conversion [...]
> ./cuda-api-wrappers/examples/modified_cuda_samples/vectorAdd/vectorAdd.cu(60): error C23
> 97: conversion from 'int' to 'cuda::grid_block_dimensions_t' requires a narrowing conversion [...]

eyalroz commented on src/cuda/api/device.hpp:
```diff
@@ -440,9 +440,9 @@ class device_t {
-		auto pci_domain_id = get_attribute(cudaDevAttrPciDomainId);
-		auto pci_bus_id = get_attribute(cudaDevAttrPciBusId);
-		auto pci_device_id = get_attribute(cudaDevAttrPciDeviceId);
+		auto pci_domain_id = get_attribute(static_cast<attribute_t>(cudaDevAttrPciDomainId));
```
Again, need to look at the `get_attribute()` function itself rather than cast its results all the time.

jvschroeder commented:
First of all sorry for creating a pull-request on the wrong branch. I changed the target branch to development.
I am not sure what will happen to your comments when I split this PR into different branches, so I'll comment on your questions / requests before I make the split you requested.

eyalroz commented:
Also, @jvschroeder : When you notice a problem, I'd like to ask you to first create an issue on the main repo's issues page. The reason is that sometimes (like in the case of some of the casts) the appropriate fix is more general than the work-around you might write. Also, some other followers might have what to say as well (which is less likely for PRs).
